# AoC 2023 Day 6
The organizer brings you over to the area where the boat races are held. The boats are much smaller than you expected - they're actually toy boats, each with a big button on top. Holding down the button charges the boat, and releasing the button allows the boat to move. Boats move faster if their button was held longer, but time spent holding the button counts against the total race time. You can only hold the button at the start of the race, and boats don't move until the button is released.

For example:

Time:      7  15   30
Distance:  9  40  200
This document describes three races:

The first race lasts 7 milliseconds. The record distance in this race is 9 millimeters.
The second race lasts 15 milliseconds. The record distance in this race is 40 millimeters.
The third race lasts 30 milliseconds. The record distance in this race is 200 millimeters.
Your toy boat has a starting speed of zero millimeters per millisecond. For each whole millisecond you spend at the beginning of the race holding down the button, the boat's speed increases by one millimeter per millisecond.

So, because the first race lasts 7 milliseconds, you only have a few options:

Don't hold the button at all (that is, hold it for 0 milliseconds) at the start of the race. The boat won't move; it will have traveled 0 millimeters by the end of the race.
Hold the button for 1 millisecond at the start of the race. Then, the boat will travel at a speed of 1 millimeter per millisecond for 6 milliseconds, reaching a total distance traveled of 6 millimeters.
Hold the button for 2 milliseconds, giving the boat a speed of 2 millimeters per millisecond. It will then get 5 milliseconds to move, reaching a total distance of 10 millimeters.
Hold the button for 3 milliseconds. After its remaining 4 milliseconds of travel time, the boat will have gone 12 millimeters.
Hold the button for 4 milliseconds. After its remaining 3 milliseconds of travel time, the boat will have gone 12 millimeters.
Hold the button for 5 milliseconds, causing the boat to travel a total of 10 millimeters.
Hold the button for 6 milliseconds, causing the boat to travel a total of 6 millimeters.
Hold the button for 7 milliseconds. That's the entire duration of the race. You never let go of the button. The boat can't move until you let go of the button. Please make sure you let go of the button so the boat gets to move. 0 millimeters.
Since the current record for this race is 9 millimeters, there are actually 4 different ways you could win: you could hold the button for 2, 3, 4, or 5 milliseconds at the start of the race.

In the second race, you could hold the button for at least 4 milliseconds and at most 11 milliseconds and beat the record, a total of 8 different ways to win.

In the third race, you could hold the button for at least 11 milliseconds and no more than 19 milliseconds and still beat the record, a total of 9 ways you could win.

To see how much margin of error you have, determine the number of ways you can beat the record in each race; in this example, if you multiply these values together, you get 288 (4 * 8 * 9).

Determine the number of ways you could beat the record in each race. What do you get if you multiply these numbers together?

**AI Prompt**

I think we can use a quadratic equation to solve the AoC Day6 Part 1 puzzle?

**AI Response**

Yes, we can definitely use a quadratic equation to solve this! Let's break down why:

For each race, we need to find how many different hold times will beat the record. Let's analyze the relationships:
- If we hold for time h milliseconds
- Race time is T milliseconds
- Time left to travel is (T-h) milliseconds
- Speed achieved is h mm/ms
- Distance traveled is h*(T-h) mm

So for a given record distance D, we need to solve:
h*(T-h) > D

This expands to:
hT - h² > D

Would you like me to help you set up this quadratic inequality for one of the example races to see how it works?

**AI Prompt**

In addition there is the constraint for h: 0 < h < T

**AI Response**

Yes, exactly! The hold time h must be:
- Greater than 0 (we must press the button for some time)
- Less than T (we can't hold longer than the race duration)

So for each race we're looking for solutions that satisfy both:
1. hT - h² > D
2. 0 < h < T

Let's try solving this for the first example race where:
- T = 7 (race time)
- D = 9 (record distance)

Would you like to solve this together?

**AI Prompt**

Yes lets solve it together

**AI Response**

Great! Let's solve it step by step.

For T=7 and D=9, we need to solve:
7h - h² > 9

Let's rearrange this into standard quadratic form:
-h² + 7h - 9 > 0

To find the boundaries of where this inequality is true, we first solve the equation:
-h² + 7h - 9 = 0

Would you like to help solve this quadratic equation? We can use the quadratic formula: (-b ± √(b² - 4ac))/(2a)
Where a=-1, b=7, and c=-9

Can you plug these values into the formula?

In [ ]:
import math
def quadratic_equation(a,b,c):
    d = b*b-4*a*c
    if d >= 0:
        x1 = (-b + math.sqrt(d))/(2*a)
        x2 = (-b - math.sqrt(d))/(2*a)
        return (x1,x2)
    else: return -1

res = quadratic_equation(-1,7,-9)
print(res)
res==-1

(1.6972243622680054, 5.302775637731995)


False

**AI Prompt**

Ok, I have two soutions for the equality

**AI Response**

Yes, you found that h ≈ 1.70 and h ≈ 5.30 are the solutions to the equality.

Since we have -h² + 7h - 9 > 0 (an inequality), the solution will be all values of h between these two points (because the parabola opens downward due to negative h²).

And remember our constraint 0 < h < 7. 

So our final solution is all integers h where:
1.70 < h < 5.30
AND
0 < h < 7

Can you list out which integer values of h satisfy both these conditions?

**AI Prompt**

How to round numbers in python, up and down?

**AI Response**

In Python, there are several ways to round numbers:

1. `math.ceil(x)` - rounds up to the nearest integer
2. `math.floor(x)` - rounds down to the nearest integer
3. `round(x)` - rounds to the nearest integer (with ties going to the nearest even number)

For example:
```python
import math
x = 1.7
print(f"Original: {x}")
print(f"Ceiling: {math.ceil(x)}")
print(f"Floor: {math.floor(x)}")
```

Would you like to try using these functions with your quadratic equation solutions?

In [ ]:
def get_ints_in_interval(x1,x2,upper_boundary):
    x1=math.ceil(x1)
    x2=min(math.floor(x2),upper_boundary)
    if x2>=x1:
        return(list(range(x1,x2+1)))
    else: return -1

print(get_ints_in_interval(1.7,5.3,7))

[2, 3, 4, 5]


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
distances = [int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
print(times)
print(distances)

[59, 70, 78, 78]
[430, 1218, 1213, 1276]


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
distances = [int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
tuples = list(zip(times,distances))
print(times)
print(distances)
print(tuples)

[59, 70, 78, 78]
[430, 1218, 1213, 1276]
[(59, 430), (70, 1218), (78, 1213), (78, 1276)]


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
distances = [-int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
tuples = list(zip(times,distances))

for b,c in tuples:
    print(b,c)

59 -430
70 -1218
78 -1213
78 -1276


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
distances = [-int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
tuples = list(zip(times,distances))

total = 1
for b,c in tuples:
    res = quadratic_equation(-1,b,c)
    if res !=-1:
        l = len(get_ints_in_interval(res[0],res[1],b))
        total *= l

print(total)

227850


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
distances = [-int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
tuples = list(zip(times,distances))

total = 1
for b,c in tuples:
    res = quadratic_equation(-1,b,c)
    if res !=-1:
        l = len(get_ints_in_interval(res[0],res[1],b))
        total *= l

print(total)

227850


In [ ]:
from aocd import get_data
inp = get_data(day=6,year=2023)
lines = inp.splitlines()
times = [int(x) for x in lines[0].split(':')[1].split(' ') if x!='']
time = int(''.join(str(x) for x in times))
distances = [int(x) for x in lines[1].split(':')[1].split(' ') if x!='']
distance = -int(''.join(str(x) for x in distances))

total = 1
b=time
c=distance
res = quadratic_equation(-1,b,c)
if res !=-1:
    l = len(get_ints_in_interval(res[0],res[1],b))
    total *= l

print(total)

42948149


## Part Two

As the race is about to start, you realize the piece of paper with race times and record distances you got earlier actually just has very bad kerning. There's really only one race - ignore the spaces between the numbers on each line.

So, the example from before:

Time:      7  15   30
Distance:  9  40  200
...now instead means this:

Time:      71530
Distance:  940200
Now, you have to figure out how many ways there are to win this single race. In this example, the race lasts for 71530 milliseconds and the record distance you need to beat is 940200 millimeters. You could hold the button anywhere from 14 to 71516 milliseconds and beat the record, a total of 71503 ways!

How many ways can you beat the record in this one much longer race?

**AI Prompt**

How to concatenate the numers in a list to one number?

**AI Response**

In Python, to concatenate numbers in a list into a single number, you can:

1. First convert each number to a string
2. Join the strings together
3. Convert back to an integer

Like this:
```python
numbers = [1, 2, 3, 4]
combined = int(''.join(str(x) for x in numbers))
```

Would you like to try this with your times and distances lists?